cfdm/data/subarray/quadraticsubarray.py

```diff
@@ -0,0 +1,132 @@
+    ...     first=(True,),
+    ...     parameters={'w': w},
+    ... )
+    >>> print(coords[...])
+    [15.0 48.75 80.0 108.75 135.0]
+    >>> coords = {{package}}.QuadraticSubarray(
+    ...     tie_points=tie_points,
+    ...     tp_indices=(slice(1, 3, None),),
+    ...     subsampled_dimensions=(0,),
+    ...     shape=(4,),
+    ...     subarea_indices=(slice(1, 2, None),),
+    ...     first=(False,),
+    ...     parameters={'w': w},
+    ... )
+    >>> print(coords[...])
+    [173.88888888888889 203.88888888888889 225.0]
+    >>> coords = {{package}}.QuadraticSubarray(
+    ...     tie_points=tie_points,
+    ...     tp_indices=(slice(3, 5, None),),
+    ...     subsampled_dimensions=(0,),
+    ...     shape=(4,),
+    ...     subarea_indices=(slice(2, 3, None),),
+    ...     first=(True,),
+    ...     parameters={'w': w},
+    ... )
+    >>> print(coords[...])
+    [255.0 289.44444444444446 319.44444444444446 345.0]
+
+
+    **Cell boundaries**
+
+    When the tie points array represents bounds tie points, then the
+    *shape* parameter describes the uncompressed bounds shape.
+
+    >>> bounds_tie_points = np.array([0, 150, 240, 240, 360])
+    >>> bounds = {{package}}.QuadraticSubarray(
+    ...     tie_points=bounds_tie_points,
+    ...     tp_indices=(slice(0, 2, None),),
+    ...     subsampled_dimensions=(0,),
+    ...     shape=(5, 2),
+    ...     subarea_indices=(slice(0, 1, None),),
+    ...     first=(True,),
+    ...     parameters={'w': w},
+    ... )
+    >>> print(bounds[...])
+    [[0.0 33.2]
+     [33.2 64.8]
+     [64.8 94.80000000000001]
+     [94.80000000000001 123.2]
+     [123.2 150.0]]
+    >>> bounds = {{package}}.QuadraticSubarray(
+    ...     tie_points=bounds_tie_points,
+    ...     tp_indices=(slice(1, 3, None),),
+    ...     subsampled_dimensions=(0,),
+    ...     shape=(3, 2),
+    ...     subarea_indices=(slice(1, 2, None),),
+    ...     first=(False,),
+    ...     parameters={'w': w},
+    ... )
+    >>> print(bounds[...])
+    [[150.0 188.88888888888889]
+     [188.88888888888889 218.88888888888889]
+     [218.88888888888889 240.0]]
+    >>> bounds = {{package}}.QuadraticSubarray(
+    ...     tie_points=bounds_tie_points,
+    ...     tp_indices=(slice(3, 5, None),),
+    ...     subsampled_dimensions=(0,),
+    ...     shape=(4, 2),
+    ...     subarea_indices=(slice(2, 3, None),),
+    ...     first=(True,),
+    ...     parameters={'w': w},
+    ... )
+    >>> print(bounds[...])
+    [[240.0 273.75]
+     [273.75 305.0]
+     [305.0 333.75]
+     [333.75 360.0]]
+
+    .. versionadded:: (cfdm) 1.9.TODO.0
+
+    """
+
+    def __getitem__(self, indices):
+        """Return a subspace of the uncompressed data.
+
+        x.__getitem__(indices) <==> x[indices]
+
+        Returns a subspace of the uncompressed data as an independent
+        numpy array.
+
+        .. versionadded:: (cfdm) 1.9.TODO.0
+
+        """
+        (d1,) = tuple(self.compressed_dimensions())
```

Comment:
This will run into a `ValueError: not enough values to unpack` if a relevant `SubsampledArray` class has been initialised without any `tie_point_indices` specified (and maybe in some other cases if they can lead to a `self.compressed_dimensions()` of `{}`), so it might be good to raise an error conveying as much at the start of the method to cover that case?

Reply:
This pattern is used a lot:
```shell
$ grep --exclude-dir test -nIR "self.compressed_dimensions()" cfdm/data/subarray
/home/david/cfdm/cfdm/data/subarray/abstract/subsampledsubarray.py:208:        subsampled_dimensions = sorted(self.compressed_dimensions())
/home/david/cfdm/cfdm/data/subarray/abstract/subsampledsubarray.py:534:        for dim in self.compressed_dimensions():
/home/david/cfdm/cfdm/data/subarray/biquadraticlatitudelongitudesubarray.py:32:        (d2, d1) = sorted(self.compressed_dimensions())
/home/david/cfdm/cfdm/data/subarray/raggedsubarray.py:28:        d1, u_dims = self.compressed_dimensions().popitem()
/home/david/cfdm/cfdm/data/subarray/bilinearsubarray.py:29:        (d2, d1) = sorted(self.compressed_dimensions())
/home/david/cfdm/cfdm/data/subarray/linearsubarray.py:29:        (d1,) = tuple(self.compressed_dimensions())
/home/david/cfdm/cfdm/data/subarray/quadraticlatitudelongitudesubarray.py:32:        (d1,) = tuple(self.compressed_dimensions())
/home/david/cfdm/cfdm/data/subarray/quadraticsubarray.py:117:        (d1,) = tuple(self.compressed_dimensions())
```
Instead of cluttering up many files, how about cda729ca9 ?

Reply:
Great idea! I've checked this works after the new commit, but for bonus points you could add some basic testing to cover the case of a SubsampledArray class without defined compressed dimensions?

Reply:
9b360d47e Can't have enough tests! turns out that  I'd put the check into `SubsampledSubarray` (where it was indeed needed) and missed out putting the  check into `CompressedArray`, where it was also needed.

Reply:
> Can't have enough tests!

Indeed. Those added in that commit look perfect, so marking this as resolved.